diff -r -u -N -x '*~' -x '*.orig' -x '*.rej' jbuilder.1.0+beta20.2-orig/src/context.ml jbuilder.1.0+beta20.2-reloc/src/context.ml
--- jbuilder.1.0+beta20.2-orig/src/context.ml	2019-12-05 11:43:40.000000000 +0100
+++ jbuilder.1.0+beta20.2-reloc/src/context.ml	2020-08-02 15:36:53.833553447 +0200
@@ -219,7 +219,7 @@
       | Some s -> Bin.parse_path s ~sep:ocamlpath_sep
     in
     let findlib_path () =
-      match kind, findlib_toolchain, Setup.library_path with
+      match kind, findlib_toolchain, Setup_reloc.library_path with
       | Default, None, Some l ->
         Fiber.return (ocamlpath @ List.map l ~f:Path.absolute)
       | _ ->
@@ -457,7 +457,7 @@
   | None   -> Path.parent t.ocaml_bin
 
 let install_ocaml_libdir t =
-  match t.kind, t.findlib_toolchain, Setup.library_destdir with
+  match t.kind, t.findlib_toolchain, Setup_reloc.library_destdir with
   | Default, None, Some d ->
     Fiber.return (Some (Path.absolute d))
   | _ ->
diff -r -u -N -x '*~' -x '*.orig' -x '*.rej' jbuilder.1.0+beta20.2-orig/src/setup_reloc.ml jbuilder.1.0+beta20.2-reloc/src/setup_reloc.ml
--- jbuilder.1.0+beta20.2-orig/src/setup_reloc.ml	1970-01-01 01:00:00.000000000 +0100
+++ jbuilder.1.0+beta20.2-reloc/src/setup_reloc.ml	2020-08-02 15:37:36.882405706 +0200
@@ -0,0 +1,9 @@
+let (library_path, library_destdir) =
+  match Sys.getenv "OPAM_SWITCH_PREFIX" with
+  | exception _ -> ( Setup.library_path, Setup.library_destdir )
+  | p ->
+    let p = String.map (fun c -> if c = '\\' then '/' else c ) p in
+    (
+      Some [ p ^ "/lib" ],
+      Some ( p ^ "/lib" )
+    )
